### PR TITLE
NOISSUE - Update Groups service to use Typescript

### LINKED ISSUE #19: Update JS SDK for groups

### PR DESCRIPTION
**Pull request by Musilah**
### What does this do?
This migrates the groups service from Javascript to Typescript.
### Which issue(s) does this PR fix/relate to?
Resolves #19 
### List any changes that modify/break current functionality
N/A
### Have you included tests for your changes?
No
### Did you document any new/modified functionality?
No

### CODE REVIEW

File: src/defs.ts [hunk 1]
```diff
@@ -44,9 +44,15 @@ export interface ThingsPage {
...
-  status?: 'enabled' | 'disabled'
+  domainID?: string
+  parentID?: string
+  description?: string
+  level?: number
+  permissions?: string[]
   createdAt?: string
   updatedAt?: string
+  status?: 'enabled' | 'disabled'
+  metadata?: Record<string, string>
```
Comment (ianmuchyri) [inline src/defs.ts:55]: you are missing param children

```
Children    []*Group  `json:"children,omitempty"`
```

File: examples/groups.ts [hunk 2]
```diff
@@ -0,0 +1,189 @@
+// Import the SDK class from the mainflux-sdk package
+import SDK from '../src/sdk'
+
+const defaultUrl = 'http://localhost'
```
Comment (dborovcanin) [inline examples/groups.ts:4]: Extract the token to the constant so that examples can be run easily.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: src/groups.ts [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+    try {
+      const response = await fetch(
+        new URL(this.groupsEndpoint, this.usersUrl).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.groupError.HandleError(errorRes.error, response.status)
+      }
+      const groupData: Group = await response.json()
```
Comment (ianmuchyri) [inline src/groups.ts:80]:
```suggestion
      const group: Group = await response.json()
```

File: src/groups.ts [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+    try {
+      const response = await fetch(
+        new URL(`${this.groupsEndpoint}/${groupID}`, this.usersUrl).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.groupError.HandleError(errorRes.error, response.status)
+      }
+      const groupData: Group = await response.json()
```
Comment (ianmuchyri) [inline src/groups.ts:114]:
```suggestion
      const group: Group = await response.json()
```

File: src/groups.ts [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+    try {
+      const response = await fetch(
+        new URL(`${this.groupsEndpoint}?${new URLSearchParams(stringParams).toString()}`, this.usersUrl).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.groupError.HandleError(errorRes.error, response.status)
+      }
+      const groupsData: GroupsPage = await response.json()
```
Comment (ianmuchyri) [inline src/groups.ts:155]:
```suggestion
      const groupsPage: GroupsPage = await response.json()
```

File: src/groups.ts [hunk 6]
```diff
@@ -0,0 +1,544 @@
...
+    try {
+      const response = await fetch(
+        new URL(`${this.groupsEndpoint}/${groupID}/permissions`, this.usersUrl).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.groupError.HandleError(errorRes.error, response.status)
+      }
+      const groupData: Permissions = await response.json()
```
Comment (ianmuchyri) [inline src/groups.ts:187]:
```suggestion
      const permissions: Permissions = await response.json()
```

Comment (ianmuchyri) [reply src/groups.ts:187]: do the same for the rest. The variables should be self descriptive

File: src/groups.ts [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+    try {
+      const response = await fetch(
+        new URL(`${this.groupsEndpoint}/${groupID}/enable`, this.usersUrl).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.groupError.HandleError(errorRes.error, response.status)
+      }
+      const enableResponse: Response = { status: response.status, message: 'Group Enabled successfully' }
```
Comment (ianmuchyri) [inline src/groups.ts:258]: `Enable` and `Disable` for group return the group

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

File: src/groups.ts [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+    try {
+      const response = await fetch(
+        new URL(`${this.groupsEndpoint}/${group.id}`, this.usersUrl).toString(),
+        options
+      )
+      if (!response.ok) {
+        const errorRes = await response.json()
+        throw this.groupError.HandleError(errorRes.error, response.status)
+      }
+      const groupData: Group = await response.json()
```
Comment (ianmuchyri) [inline src/groups.ts:226]:
```suggestion
      const group: Group = await response.json()
```

Comment (ianmuchyri) [reply src/groups.ts:226]: check the rest in the page that are not changed

**Review by ianmuchyri — CHANGES_REQUESTED** [review]

**Review by ianmuchyri — APPROVED** [review]
LGTM

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**